pkg/adaptor/elasticsearch/elasticsearch.go

```diff
@@ -77,69 +134,124 @@ func init() {
+	return errors.New("Start is unsupported for elasticsearch")
 }
 
 // Listen starts the listener
 func (e *Elasticsearch) Listen() error {
-	e.setupClient()
-	e.indexer.Start()
-	e.running = true
-
-	go func(cherr chan *elastigo.ErrorBuffer) {
-		for err := range e.indexer.ErrorChannel {
-			e.pipe.Err <- adaptor.NewError(adaptor.CRITICAL, e.path, fmt.Sprintf("elasticsearch error (%s)", err.Err), nil)
-		}
-	}(e.indexer.ErrorChannel)
-
+	log.With("path", e.path).Infoln("adaptor Listening...")
 	defer func() {
-		if e.running {
-			e.running = false
-			e.pipe.Stop()
-			e.indexer.Stop()
-		}
+		log.With("path", e.path).Infoln("adaptor Listen closing...")
+		e.pipe.Stop()
 	}()
 
 	return e.pipe.Listen(e.applyOp, e.typeMatch)
 }
 
 // Stop the adaptor
 func (e *Elasticsearch) Stop() error {
-	if e.running {
-		e.running = false
-		e.pipe.Stop()
-		e.indexer.Stop()
-	}
+	log.With("path", e.path).Infoln("adaptor Stopping...")
+	e.pipe.Stop()
+
+	close(e.doneChannel)
+	e.wg.Wait()
+
+	log.With("path", e.path).Infoln("adaptor Stopped")
 	return nil
 }
 
 func (e *Elasticsearch) applyOp(msg message.Msg) (message.Msg, error) {
-	m, err := message.Exec(message.MustUseAdaptor("elasticsearch").(elasticsearch.Adaptor).UseIndexer(e.indexer).UseIndex(e.index), msg)
+	_, msgColl, _ := message.SplitNamespace(msg)
+	err := e.client.Write(message.From(msg.OP(), e.computeNamespace(msgColl), msg.Data()))(nil)
+
 	if err != nil {
-		e.pipe.Err <- adaptor.NewError(adaptor.ERROR, e.path, fmt.Sprintf("elasticsearch error (%s)", err), msg.Data)
+		e.pipe.Err <- adaptor.NewError(adaptor.ERROR, e.path, fmt.Sprintf("write message error (%s)", err), msg.Data)
 	}
-	return m, err
+	return msg, err
+}
+
+func (e *Elasticsearch) computeNamespace(Type string) string {
+	return fmt.Sprintf("%s.%s", e.index, Type)
 }
 
-func (e *Elasticsearch) setupClient() {
-	// set up the client, we need host(s), port, username, password, and scheme
-	client := elastigo.NewConn()
+func (e *Elasticsearch) setupClient(conf Config) error {
+	uri, err := url.Parse(conf.URI)
+	if err != nil {
+		return InvalidURIError{conf.URI}
+	}
+
+	hostsAndPorts := strings.Split(uri.Host, ",")
+	stringVersion, err := determineVersion(fmt.Sprintf("%s://%s", uri.Scheme, hostsAndPorts[0]))
+	if err != nil {
+		return err
+	}
+
+	v, err := version.NewVersion(stringVersion)
+	if err != nil {
+		return VersionError{conf.URI, stringVersion, err.Error()}
+	}
 
-	if e.uri.User != nil {
-		client.Username = e.uri.User.Username()
-		if password, set := e.uri.User.Password(); set {
-			client.Password = password
+	httpClient := http.DefaultClient
+	if conf.Timeout != "" {
+		t, err := time.ParseDuration(conf.Timeout)
+		if err != nil {
+			return InvalidTimeoutError{conf.Timeout}
+		}
+		httpClient = &http.Client{
+			Timeout: t,
 		}
 	}
 
-	// we might have a port in the host bit
-	hostBits := strings.Split(e.uri.Host, ":")
-	if len(hostBits) > 1 {
-		client.SetPort(hostBits[1])
+	for _, vc := range clients.Clients {
```

Comment:
If we make the registry generic, I think a function to return the proper client given the adaptor type and version would be better than having this logic in the adaptor. Even for just elasticsearch, I think it would be better to have this in the registry.